### DISCUSSION TITLE
Micropython for the BL702 from Bouffalo Lab?

### DISCUSSION THREAD

**Discussion by rog77** [micropython/micropython#10667] [Other]
Hi,

I have had a look but can't see if this exists or not, any help would be greatly appreciated.

Does anyone know if there is a port of micropython for the BL702? If not, is it very difficult to create one?

I would really like to be able to use the BL702 that is on the Tang Nano FPGA series for some custom functionality that is already written in micropython. 

Thanks in advance for any advice!

**Comment by rog77** [comment]
So I read the core on that chip is an RV32, so it certainly sounds like it should be possible. I will keep looking, but am more than happy to hear suggestions or advice.

**Comment by mattytrentini** [comment]
There is no port that I'm aware of. But I have interest in the BL702 - and, indeed, other Bouffalo Lab chips, particularly the BL808. The BL808 is used in the [Pine Ox64](https://wiki.pine64.org/wiki/Ox64) and the [Sipeed M1s](https://wiki.sipeed.com/hardware/en/maix/m1s/m1s_module.html) and the the BL702 is used in the [Sipeed M0sense](https://wiki.sipeed.com/hardware/en/maixzero/sense/maix_zero_sense.html). 

Bouffalo Labs provide the [bl_mcu_sdk](https://github.com/bouffalolab/bl_mcu_sdk) which appears to be reasonably well structured and supported, though still somewhat in flux. There are some question marks around their radio control; it seems that firmware blobs may be required for BLE and Wifi, though there's no official word yet. However, most other core micro features seem more mature.

Creating a MicroPython port takes significant effort - you have to spend a lot of time learning the manufacturer's SDK. But I think the Bouffalo Labs range of micros would an excellent platform to run MicroPython on.

**Reply by rog77** [reply]
I guess a question is: where to start?
Bare metal, or FreeRTOS, or even embedded?
What is the closest already existing mcu to try to copy?

I have the SDK up and running (sipeed version) , and decided to pick on my RV-Debugger, I managed to build and load some examples that didn't do much, but then rebuilt the sipeed firmware, and get it operational again.

**Comment by mattytrentini** [comment]
Bare-metal, following the [Porting MicroPython](https://docs.micropython.org/en/latest/develop/porting.html) documentation would be my approach. 

First milestone would be to get _just enough_ of the port together so that the REPL can be accessed over a UART (or USB CDC if it's not too difficult). The [examples folder](https://github.com/bouffalolab/bl_mcu_sdk/tree/master/examples) contains a lot of useful guidance. After the REPL is available, start adding Pins and the other peripherals. 

All of the ports are - by nature - quite different, but I would suggest looking to the [rp2 port](https://github.com/micropython/micropython/tree/master/ports/rp2) for inspiration, since it's one of the most recent, quite clean and _relatively_ straightforward (particularly since it covers just one micro).

**Reply by rog77** [reply]
Thanks for the advice, I am not a C programmer so expect this to be very painful...

Speaking of the RV-Debugger, the firmware has dual UART and JTAG, exposing two COM ports bect to the desktop. For my purposes, I would need to find a way of keeping the port that does JTAG exposed, presumably though micropython, as well as probably needing uart on both the USB side and via the GPIO, does this sound achievable?
